libraries/chain/include/eosio/chain/finalizer_set.hpp

```diff
@@ -1,59 +1,49 @@
+      finalizer_set& operator=(finalizer_set&&);
 
-      std::string  description;
-      uint64_t     fweight = 0; // weight that this finalizer's vote has for meeting fthreshold
-      fc::crypto::blslib::bls_public_key  public_key;
+      uint32_t                                       generation = 0; ///< sequentially incrementing version number
+      uint64_t                                       fthreshold = 0;  ///< vote fweight threshold to finalize blocks
+      std::vector<finalizer_authority>               finalizers; ///< Instant Finality voter set
 
-      friend bool operator == ( const finalizer_authority& lhs, const finalizer_authority& rhs ) {
-         return tie( lhs.description, lhs.fweight, lhs.public_key ) == tie( rhs.description, rhs.fweight, rhs.public_key );
-      }
-      friend bool operator != ( const finalizer_authority& lhs, const finalizer_authority& rhs ) {
-         return !(lhs == rhs);
-      }
+      auto operator<=>(const finalizer_set&) const;
    };
 
-   struct finalizer_set {
-      finalizer_set() = default;
-
-      finalizer_set( uint32_t version, uint64_t fthreshold, std::initializer_list<finalizer_authority> finalizers )
-      :version(version)
-      ,fthreshold(fthreshold)
-      ,finalizers(finalizers)
-      {}
-
-      uint32_t                                       version = 0; ///< sequentially incrementing version number
-      uint64_t                                       fthreshold = 0;  // vote fweight threshold to finalize blocks
-      vector<finalizer_authority>                    finalizers; // Instant Finality voter set
-
-      friend bool operator == ( const finalizer_set& a, const finalizer_set& b )
-      {
-         if( a.version != b.version ) return false;
-         if( a.fthreshold != b.fthreshold ) return false;
-         if ( a.finalizers.size() != b.finalizers.size() ) return false;
-         for( uint32_t i = 0; i < a.finalizers.size(); ++i )
-            if( ! (a.finalizers[i] == b.finalizers[i]) ) return false;
-         return true;
-      }
-
-      friend bool operator != ( const finalizer_set& a, const finalizer_set& b )
-      {
-         return !(a==b);
-      }
+   using finalizer_set_ptr = std::shared_ptr<finalizer_set>;
+
+   /**
+    * Block Header Extension Compatibility
+    */
+   struct hs_finalizer_set_extension : finalizer_set {
+
+      static constexpr uint16_t extension_id() { return 2; } // TODO 3 instead?
+      static constexpr bool     enforce_unique() { return true; }
+
+      hs_finalizer_set_extension() = default;
+      hs_finalizer_set_extension(const hs_finalizer_set_extension&) = default;
+      hs_finalizer_set_extension( hs_finalizer_set_extension&& ) = default;
+
+      hs_finalizer_set_extension& operator=(const hs_finalizer_set_extension&) = default;
+      hs_finalizer_set_extension& operator=(hs_finalizer_set_extension&&) = default;
+
+      hs_finalizer_set_extension(const finalizer_set& s);
```

Comment:
I think we should remove all these and the implementation of `  hs_finalizer_set_extension(const finalizer_set& s)`